src/validation.cpp

```diff
@@ -1398,6 +1398,12 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
             // Provide the wtxid of the mempool tx so that the caller can look it up in the mempool.
             results_final.emplace(wtxid, MempoolAcceptResult::MempoolTxDifferentWitness(iter.value()->GetTx().GetWitnessHash()));
         } else {
+            if (wtxid == child->GetWitnessHash() && !quit_early) {
+                Assume(tx == package.back());
+                txns_package_eval.push_back(tx);
```

Comment:
db6eed0cdfb8a90b54fee4955fd9b07d5fc1817b
Consider everything, including the child, to be valid.
Previously, the execution would terminate at `txns_package_eval.empty()` gate after the loop.
After this commit, this won't pass — and everything would be validated once again.

Reply:
I think it's the opposite - we avoid re-validating things now.

> Consider everything, including the child, to be valid.

Ok, we can break this down. Let's split it into 2 types of scenarios: non-CPFP case and CPFP case. In the CPFP case, the parent needs the child in order to pass.

(1) Non-CPFP case
(1a) Before this commit
Inside this loop, **validate** parent. The parent passes and is submitted to the mempool.
Then, **validate** child. The child passes and is submitted.
After the loop, we quit because `txns_package_eval.empty()`.
How many times we validated something: 2

(1b) After this commit
Inside this loop, **validate** parent. The parent passes and is submitted to the mempool.
Inside this loop, at this condition, `txns_package_eval.push_back(child)` and break.
After the loop, we **validate** using `AcceptPackageWrappingSingle({child})`.
How many times we validated something: 2

(2) CPFP case
(2a) Before this commit
Inside loop, **validate** parent. The parent fails due to too low fee. `txns_package_eval.push_back(parent)`.
Inside this loop, **validate** child. The child fails due to missing inputs. `txns_package_eval.push_back(child)`.
After the loop, we **validate** using `AcceptPackageWrappingSingle({parent, child})`.
How many times we validated something: 3

(2b) After this commit
Inside loop, **validate** parent. The parent fails due to too low fee. `txns_package_eval.push_back(parent)`.
Inside this loop, at this condition, `txns_package_eval.push_back(child)` and break.
After the loop, we **validate** using `AcceptPackageWrappingSingle({parent, child})`.
How many times we validated something: 2

Reply:
Thank you for helping me understand this. I understand you are right now.